File: src/tests/webp.zig
const std = @import("std");
const generateGradientPixels = @import("shared.zig").generateGradientPixels;

test "webp encode" {
    const c = @cImport({
        @cInclude("webp/encode.h");
    });

    const width: c_int = 16;
    const height: c_int = 16;

    const rgba_pixels = try generateGradientPixels(std.testing.allocator, @intCast(width), @intCast(height), true);
    defer std.testing.allocator.free(rgba_pixels);

    var webp_buffer: ?[*]u8 = null;
    const webp_size = c.WebPEncodeRGBA(rgba_pixels.ptr, width, height, width * 4, 90, &webp_buffer);
    try std.testing.expect(webp_size > 0);
    defer c.WebPFree(webp_buffer);
}

test "webp decode" {
    const c = @cImport({
        @cInclude("webp/decode.h");
        @cInclude("webp/encode.h");
    });

    const rgba_pixels = @embedFile("test-images/orange.webp");

    var decoded_width: c_int = 0;
    var decoded_height: c_int = 0;
    const decoded_pixels = c.WebPDecodeRGBA(rgba_pixels.ptr, rgba_pixels.len, &decoded_width, &decoded_height);
    try std.testing.expect(decoded_pixels != null);
    defer c.WebPFree(decoded_pixels);
}

File: src/tests/webp.c
#include <stdio.h>
#include <stdlib.h>

#include "shared.h"
#include <webp/decode.h>
#include <webp/encode.h>

bool test_webp_encode() {
  const int width = 16;
  const int height = 16;

  auto rgba_pixels = generate_gradient_pixels(width, height, true);

  uint8_t *webp_buffer = NULL;
  size_t webp_size = WebPEncodeRGBA(rgba_pixels.data, width, height, width * 4,
                                    90, &webp_buffer);

  free(rgba_pixels.data);
  WebPFree(webp_buffer);
  return webp_size > 0;
}

bool test_webp_decode() {
  auto rgba_pixels = read_file("src/tests/test-images/orange.webp");
  if (!rgba_pixels.data)
    return false;

  int decoded_width, decoded_height;
  uint8_t *decoded_pixels = WebPDecodeRGBA(rgba_pixels.data, rgba_pixels.len,
                                           &decoded_width, &decoded_height);

  free(rgba_pixels.data);
  if (!decoded_pixels)
    return false;
  WebPFree(decoded_pixels);
  return true;
}

int main() {
  if (!test_webp_encode()) {
    printf("webp_encode failed\n");
    return 1;
  }

  if (!test_webp_decode()) {
    printf("webp_decode failed\n");
    return 1;
  }

  return 0;
}
